autosklearn/ensemble_building/builder.py

```diff
@@ -0,0 +1,925 @@
+            if not any(candidates):
+                best = sorted(to_delete, key=lambda r: (r.loss, r.num_run))[0]
+                candidates = [best]
+                to_delete.remove(best)
+
+            if any(to_delete):
+                self.logger.info(
+                    f"Deleting runs {to_delete} due to"
+                    f" max_models={self.max_models_on_disc} and/or"
+                    f" memory_limit={self.model_memory_limit}"
+                )
+                self.delete_runs(to_delete)
+
+        # If there are any candidates, perform candidates selection
+        if any(candidates):
+            candidates, to_delete = self.candidate_selection(
+                runs=candidates,
+                dummies=dummies,
+                better_than_dummy=True,
+                nbest=self.ensemble_nbest,
+                performance_range_threshold=self.performance_range_threshold,
+            )
+            if any(to_delete):
+                self.logger.info(
+                    f"Deleting runs {to_delete} due to"
+                    f" nbest={self.ensemble_nbest} and/or"
+                    f" performance_range_threshold={self.performance_range_threshold}"
+                )
+                self.delete_runs(to_delete)
+        else:
+            candidates = dummies
+            self.logger.warning("No real runs to build ensemble from")
+
+        # If there's an intersect with models that have some predictions on the
+        # test subset, use that subset, otherwise use all of the candidates
+        candidates_set = set(candidates)
+        test_subset = {r for r in candidates if r.pred_path("test").exists()}
+
+        if len(test_subset) > 0:
+            candidates = sorted(test_subset, key=lambda r: r.id)
+            test_models = candidates
+
+            to_delete = candidates_set - test_subset
+            if any(to_delete):
+                self.logger.info(
+                    f"Deleting runs {to_delete} due to runs not"
+                    ' having "test_predictions" while others do not:'
+                    f"\nHave test_predictions = {test_subset}"
+                    f"\nNo test_predictions = {to_delete}"
+                )
+                self.delete_runs(candidates_set - test_subset)
+
+        else:
+            candidates = sorted(candidates_set, key=lambda r: r.id)
+            test_models = []
+
+        # Save the candidates for the next round
+        with self.previous_candidates_path.open("wb") as f:
+            pickle.dump({run.id: run for run in candidates}, f)
+
+        # If there was any change from the previous run, either in terms of
+        # runs or one of those runs had its loss updated, then we need to
+        # fit the ensemble builder
+        previous_candidate_ids = set(previous_candidates)
+        current_candidate_ids = set(run.id for run in candidates)
+        difference = previous_candidate_ids ^ current_candidate_ids
```

Comment:
I don't understand this. Doesn't the variable `previous_candidate_ids` include all candidates available on disk, and `current_candidate_ids` all candidates we consider? When can this be not empty?

Reply:
Sorry, maybe the variable names should be better or at least documented better.

* `previous_candidate_ids` are runs that were used in the ensemble fitting in the previous iteration.
* `current_candidate_ids` are runs that are considered for ensemble fitting this iteration. This can overlap with `previous_candidates_ids`. 

`difference` would only be empty when all the candidates for ensemble fitting in this iteration are exactly the same as the ones as last iteration. i.e. there is no change in the runs passed to `ensemble.fit()` from the previous time the builder was run.

This coupled with `was_updated_candidates` is what replaces the hashing mechanism. If all the runs are the same and they were not updated, then there is no need to fit a new ensemble as nothing has changed.

Reply:
Thanks for the explanation. With this, and looking at the code a 2nd time, I can follow the logic.